identityfederation.go

```diff
@@ -0,0 +1,184 @@
+	ClientID string
+	// IDToken is the identity generated by the identity provider for the workflow.
+	// Either IDToken or IDTokenGenerator must be provided, but not both.
+	IDToken string
+	// IDTokenGenerator is an optional function that dynamically generates a fresh ID token.
+	// This is useful for long-running applications where ID tokens may expire.
+	// Either IDToken or IDTokenGenerator must be provided, but not both.
+	IDTokenGenerator func() (string, error)
+	// BaseURL is an optional base URL for the API server to which we'll connect. Defaults to https://api.tailscale.com.
+	BaseURL string
+}
+
+// TokenExchangeResponse represents the response from the Tailscale token exchange endpoint.
+type TokenExchangeResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"` // in seconds
+	Scope       string `json:"scope"`
+}
+
+// jwtClaims represents the claims in a JWT token (minimal set for validation).
+type jwtClaims struct {
+	Exp int64 `json:"exp"`
+}
+
+// tokenTransport implements http.RoundTripper and handles token exchange and caching.
+type tokenTransport struct {
+	config    IdentityFederationConfig
+	transport http.RoundTripper
+	token     string
+	expiresAt time.Time
+}
+
+// HTTPClient constructs an HTTP client that authenticates using identity federation.
+// The client automatically handles token exchange and caching.
+func (c IdentityFederationConfig) HTTPClient() (*http.Client, error) {
+	if c.ClientID == "" {
+		return nil, fmt.Errorf("ClientID is required")
+	}
+	if c.IDToken != "" && c.IDTokenGenerator != nil {
+		return nil, fmt.Errorf("IDToken and IDTokenGenerator cannot both be provided")
+	}
+	if c.IDToken == "" && c.IDTokenGenerator == nil {
+		return nil, fmt.Errorf("either IDToken or IDTokenGenerator must be provided")
+	}
+	if c.BaseURL == "" {
+		c.BaseURL = defaultBaseURL.String()
+	}
+
+	transport := &tokenTransport{
+		config:    c,
+		transport: http.DefaultTransport,
+	}
+
+	// Perform initial token exchange to validate configuration early
+	if err := transport.refreshToken(context.Background()); err != nil {
+		return nil, err
+	}
+
+	return &http.Client{
+		Transport: transport,
+		Timeout:   defaultHttpClientTimeout,
+	}, nil
+}
+
+// RoundTrip executes a single HTTP transaction, refreshing the API access token if necessary.
+func (t *tokenTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if t.token == "" || time.Now().After(t.expiresAt) {
+		if err := t.refreshToken(req.Context()); err != nil {
+			return nil, err
+		}
+	}
+
+	reqClone := req.Clone(req.Context())
+	reqClone.SetBasicAuth(t.token, "")
+
+	return t.transport.RoundTrip(reqClone)
+}
+
+// refreshToken performs the token exchange and updates the cached token.
+func (t *tokenTransport) refreshToken(ctx context.Context) error {
+	if t.config.IDTokenGenerator != nil && validateIDToken(t.config.IDToken) != nil {
```

Comment:
Your suggestion of just having one `IDTokenGenerator` parameter on the config sounds good to me, as it gets rid of a lot of these if/else branches.

At that point, you could also just call it `IDToken` again.